src/lib/core/popen.c

```diff
@@ -229,6 +230,8 @@ handle_new(struct popen_opts *opts)
 	for (i = 0; i < lengthof(handle->ios); i++)
 		iostream_clear(&handle->ios[i]);
 
+	fiber_cond_create(&handle->completion_cond);
+
```

Comment:
Nit: Matching `fiber_cond_destroy` is missing. (I know it does nothing, but it wouldn't do any harm to put in `handle_free`).

Reply:
It is really good point (I totally missed it, shame on me).

`<popen handle>:close()` should unblock `<popen handle>:wait()` -- and it should be tested and documented. Otherwise the `:wait()` may stuck forever.

`fiber_cond_destroy()` requires the cond to have no waiters and I'll add `fiber_cond_broadcast()`, which frees the list and wake ups all the fibers.

I'll adjust `popen_wait_timeout()` condition to take care to the handle's pre-destroy state. I would prefer to don't complicate `handle_free()` with waiting and allow `popen_wait_timeout()` to detect handle's closing and don't read freed memory, but I don't know ATM how to do so. Will investigate.

Reply:
It seems, I can use `popen_find()` (before any access to the handle) to eliminate access to the freed handle.

Reply:
Updated. The code becomes a bit tricky now (see comments in the commit message and in the code). If you have suggestions how to make it better, I'm open.

Reply:
Your comment lead me to look more around: I found #7653 and fixed it in the same patch (those things look tightly connected).